README

=pod

=head1 README

You can use perldoc to read this document: C<perldoc README>.

=head1 DESCRIPTION

=head2 Verify

Before you do anything with the downloaded package, verify the signature.

Download our GnuPG key from http://www.bsi.bund.de/produkte/dicop/download.htm
and import it into GnuPG.

=over 2

=item *

Verify the signature of the .tar.gz file by downloading the .tar.gz.asc file
and run 

	gpg Dicop-Workerframe-2.19.tar.gz.asc

Replace "2.19" with the actual version number you got.

=back

If the signature does not verify ok, please notify us immediately. 
B<DO NOT> run perl Makefile.PL or anything else nor install the package when
the signature cannot be verified ok. B<Your system might get compromised!>

=head2 File Layout

These files are in the Dicop-Workerframe package:

	doc/			additional documentation

	src/			source for the framework
	src/main/		main framework source
	src/include/		some include files for src/main/
	src/include/encoding/	auto-generated tables for encoding and decoding

	src/test/		an example (test) worker
	src/pwdtest/		(single pwd at a time) testworker for the testsuite
	src/extract/		example test worker for extraction from image files
	src/test_par/		(pwds in batches) testing the parallel interface
	src/test_conv/		test charset conversion
	src/test_enc/		test encodings and their conversion
	src/test_sbc/		test single-byte encodings

	lib/			libdicop is build and put into this directory

	build/			files for package management and build process
	t/			testsuite
	t/test-worker		test workers are created in this directory
	t/target/		test data for the testsuite
	t/output/		expected output for the testsuite

=head2 Short Overview

This is a framework for generating so-called workers (hence the name
workerframe) for the DiCoP project. The resulting executables are able to be
run standalone, but their real purpose is to be run in an DiCoP cluster
environment.

So you should install the DiCoP server and client first. See
L<http://www.bsi.bund.de/> and search for "dicop", then download and install
the server and at least one client.

This framework provides all the necessary routines to get an "empty" worker
done, e.g. something that just walks a bit of a password/key space and
applies an "empty check" (e.g. none at all) to each password/key.

Workers are usually called by the DiCoP client and take their arguments from
the command line and/or additional files. They print their results to
STDOUT, which is then collected by the client.

All the initialization is done by the framework. The framework contains a
password generator (actually more than one), and this generator cycles through
all passwords in the given range and calls a check function (called
C<dofunction>) for each password.

The framework can walk password/key rooms that are described by different
types of character sets. For details please see C<doc/API.pod>.

The idea behind this to keep the system simple, the communication
human-readable (thus easily debugable) and avoid dependencies as much as
possible. The workers should (and do) be able to run as simple programs on a
text-console, with no dependencies except for the c library. This makes it
easier to deploy them on systems with quite different operating systems or
hardware.

To simplify the development of workers further, the common parts were
relocated into the workerframe. This frame manages all the boring things
like the interaction between client and worker, the setup of the password
generator (which is the thing that know how to walk the key room from C<start>
to C<end>) and then the actual generating of all keys in the key room.

There are only a couple of functions you need to fill in. Please refer
to C<doc/API.pod> for a listing and their meaning.

Have a look into the C<src/pwdtest> and C<src/test> directory, they contains
two sample workers. Also noteworthy are the C<test.pl> and C<prime.pl> test
workers distributed with the DiCoP project.

That's all, folks!

=head1 INSTALLATION

=head2 Testing

The worker framework does not need to be installed, nor to be copied into
the server/client. It is just there to build new workers for the cluster.

Just untar/ungzip the package to a directory of your choice:

	tar -xzf Dicop-Workerframe-2.19.tar.gz

Replace "2.19" by the actual version you got.

        perl Makefile.PL
        make test

to run the testsuite. It should pass all the tests. If not, please notify us.

=head3 Windows

Under mswin32, replace C<make> by C<nmake>:

        perl Makefile.PL
        nmake test

=head2 Install

You do B<NOT> need to C<make install>. There is nothing to install.

Files that are compiled with the workerframe need to be copied into the
servers C<worker/> dir, where they can be picked up by the clients.

They will use a file for the charsets, this file is written by the server
whenever the charsets change. The default name for that file is charsets.def
and it is located in the C<worker/> dir. Each worker will try to find this file
from it's own directory, and if this fails, one directory up. Here is a sample
layout:

	worker/charsets.def		generated by server
	worker/linux/test
	worker/os2/test			etc, pp.

All workers will use the same charsets.def file, which is generated by
the server and distributed to the clients automatically.

A sample charsets.def file (used by the testsuite) is included with this
package, you can find it in C<t/test-worker/charsets.def>.

=head1 DEVELOPMENT

To develop a new worker, have a look at the two sample workers included in
the distribution: C<test> and C<pwdtest>. Anything that is in these two
directories (C<src/test> and C<src/pwdtest>) is needed to build a new worker
- nothing more or less ;) See also L<DESCRIPTION|DESCRIPTION>.

=head1 Supported platforms

The latest version of this framework was tested the following platforms:

=head2 Linux with GCC

We tested with GCC 3.3.5 and GCC 4.1.0.

=head2 Windows with MS Visual C++

The project should compile cleanly under a modern Windows using Microsoft
Visual C++ 2003. You need 64 bit support for file pointers, e.g.
C<fseeki64> and C<ftelloi64> must exist.

When compiling under Microsoft Visual C++ 8, you need to specify

        /NODEFAULTLIB:libc

in the linker flags, or it will complain. *sigh*

=head2 Windows Vista

Windows Vista restricts non-Win32 apps to 32 Mbyte memory. We have not
yet determined whether our workers are hit by this limitation. The
following page gives background information and workarounds:

	http://www.trnicely.net/misc/vista.html

=head2 Others

It should work on other platforms, too. However, we can not guaranty that
it will work on all platforms. If you experience any problems, please contact
us (see below).

=head1 AUTHOR

(C) Bundesamt fuer Sicherheit in der Informationstechnik 1998-2007.

For licensing information please refer to the LICENSE file.

=head2 Contact

Address: BSI
	 Referat 123
	 Godesberger Allee 185-189
	 53175 Bonn
	 Germany
email:	 dicop @ bsi.bund.de	(for public key see dicop.asc or web site)



CHANGES

=pod

=head1 History revision for Dicop-WorkerFrame

This document details all the changes from v2.15 onwards - for an older
change log please see the C<CHANGES-2.14> file.

=head2 2008-10-01 (v2.21 rev 698)

=over 2

=item added

	convert from single-byte encodings to ASCII
	query locale info before --identify, to enable proper output

=back

=head2 2008-06-11 (v2.20 rev 615)

=over 2

=item fixed

	update to Unicode 5.1.0 (w/o uppercase ß)
	small doc fixes in doc/Config.pod and others
	More fixes to try and get it compile on Solaris

=item added

	pwdgen_truncate_string(pwd, str, X) - truncate string to X characters
	conversion from UTF-16/UTF-32 to UTF-8

=back

=head2 2007-11-26 (v2.19 rev 580)

=over 2

=item fixed

	renamed "prefix_encoding" to "config_encoding"
	document all the new config switches and generally update doc/Config.pod
	fix segfault in the extract code, when having either one encoding
	  distinct from the input encoding, or simple more than one encoding
	fix for darwin builds
	better LC_TYPE parsing
	try to not include err.h on Solaris (untested)
	require Perl 5.6.2 for the build stage
	add support for CP863

=back

=head2 2007-09-14 (v2.18 rev 546)

Please note that the API has changed slightly, you must adapt your code!

Instead of C<< pws->pwd >> and C<< pwd->length >>, use
C<< pwd->cur->content >> and C<< pwd->cur->bytes >> to access the current
password. See C<< doc/I18N.pod >> for more details.

=over 2

=item added

	support for "prefix" and "prefix_encoding" in config files
	  (unlike "password_prefix", "prefix" is in ASCII/UTF_8 etc)
	pwdgen_strdup(), and pwdgen_strrev()
	pwdgen_uc_first(), pwdgen_uc_last() etc.

=item fixed

	dictionary code can also try the password exactly as written

=back

=head2 2007-08-30 (v2.18 rev 525)

=over 2

=item added

	convert strings from/to different character sets
	support for different encodings (UTF-8, ISO 8859-1 etc)
	added a test_enc worker that tests encodings and conversions
	--identify command line switch support
	allow "dictionary" in config files as alias for "dictionary_file"

=item fixed

	Fixed a lot of compiler warnings, and made the compiler options
	  detection much more robust, so it will build on more platforms
	Removed the unused "image_type" from config file support and code
	extract code uses a bit more memory, and checks for timeouts now

=back

=head2 2006-12-20 (v2.17 rev 385)

Too many things have changed to be listed here, so this is just a short
overview. Please note that the API has changed slightly, you must adapt
you code to the new function calling style!

See the example in src/test/test.c for guidance.

=over 2

=item fixed

	build a real Makefile in Makefile.PL
	remove needless "mi" and "mg" files
	remove broken Umlaut handling in lc(), uc(), a2h() and h2a()
	fix compilation to work again after shuffling of include files
	convert all *.c and *.h to unix line endings again
        remove needless double includes, fix prototypes of a2h() and h2a()
	sprinkle some const qualifiers
	fix the unsigned vs. signed warnings from compilers
	define handy macros for libtomcrypt support in dicop.h
	define macros for file support (read/open(close) with error handling
	very large file support (> 2GB)
	remove the outdated pwdblock worker example
	output the status of the work in progress every couple of minutes
	add Doxygen support
	16675 tests

=back

=head2 2006-05-02 (v2.16)

=over 2

=item fixed

	fix dict.t failing due to machines being to fast
	fix (C) messages and contact info
	add SIGNATURE for CPAN
	remove a few needless #include statements
	remove needless double change lists in source files
	fix a few unsigned vs. signed warnings from gcc
	fix t/extract2.t to work properly
	16986 tests

=back

=head2 2005-04-20 (v2.15 build 3)

=over 2

=item fixed

	forgot to include t/par.t into MANIFEST
	error message about number of arguments
	add "const" to some arguments that are not supposed to change
	wcr src: Total: lines 6098, words 24929, chars 171821
	16334 tests

=back

=head2 2005-03-17 (v2.15 build 2)

=over 2

=item fixed

	mark src/pwdblock as outdated
	main.c: supports now parallel (pwds in blocks) interface

=item added

	src/main/hexdump.c and src/include/hexdump.h
	src/test_par - parallel interface
	wcr src: Total: lines 6088, words 24861, chars 171364
	16334 tests

=back

=head2 2005-03-16 (v2.15 build 1)

=over 2

=item added

	add a blurb to each .c file in src/main about it's contents
	fix (c) years to 2005
	doc/Parallel.pod
	src/main/pwd_par.c
	16024 tests

=back

=head1 AUTHOR

(c) Bundesamt fuer Sicherheit in der Informationstechnik 1998-2008

DiCoP is free software; you can redistribute it and/or modify it under the
terms of the GNU General Public License version 2 as published by the Free
Software Foundation.

See the file LICENSE or L<http://www.bsi.de/> for more information.

=cut



src/main/pwdread.c
/*!
 * @file
 * @ingroup workerframe
 * @brief Implements the various config file reading routines.
 * 
 * @copydoc copyrighttext
*/

#include <dicop.h>
#include <pwdread.h>
#include <pwd_util.h>

/* ************************************************************************* */
/* INTERNAL: read file with charset definitions and build character sets from
             them, inclusive grouped charsets from the simple ones
   return codes: ptr to charsets - okay
		 NULL 		 - error 	*/

/* defined in pwdgen.c */
extern unsigned int MAXSETS;

/* Thou Shallst Not Add Arbitrary Limits To Thy Programme - but I did ... */

/** Buffer size for reading character sets */
#define BUFSIZE 16384
/** Length of one field in a charset definition */
#define MAXFIELDSIZE 16

struct ssCharset *
pwdgen_read_charsets (void)
{
  int linenr, convert, mul, stage, s, idx;
  unsigned int pos, set;
  FILE *file;
  unsigned int k, j, i, count;
  unsigned char line[BUFSIZE + 16];	/* input lines up to BUFSIZE characters */
  unsigned char buffer[16];	/* temp. buffer for conversations */
  unsigned char temp[BUFSIZE];	/* temp. buffer for conversations */
  unsigned int temp_values[8];	/* temp. buffer for values */
  struct ssCharset *charset;	/* to store read in charsets */
  int alloc = 0;		/* already allocated memory? */

  printf ("\n Opening global charset description file...");

  if (NULL == (file = pwdgen_find_file ("charsets.def", "r")))
    {
      return NULL;
    }

  MAXSETS = 0;

  /* read in line by line, format is as follows:
     count=6
     # comments
     0:1:41:42...       simple charset, fields:
                        type id  chars
                        0    1.. 41 42 43 ...
     1:2:1=1:2=1...     grouped charset, fields: 
                        type id  pairs (position 1..length or -1..-length)
                        1    1.. position=charset
     2:2:0,32,1,2:...   dictionary, fields: 
                        type id  pairs (pos= 0|1) 
                        2    1.. pos,set,startlen,endlen: 
   */
  charset = NULL;		/* silence the might be used unitialized warning */
  linenr = -1;
  set = 0;			/* index of set we are reading in */
  while (!feof (file))
    {
      linenr++;
      count = 0;
      line[0] = 0;
      while (count < BUFSIZE && !feof (file))
	{
	  line[count] = (unsigned char)fgetc (file);
	  /* handles only 0xd0xa or 0xa files, not 0xd (like MAC) */
	  if (line[count] == 0x0d || line[count] == 0x0a)
	    {
	      line[count] = 0;
	      break;		/* kill line end/ line feed */
	    }
	  line[++count] = 0;
	}
      if (count >= BUFSIZE)
	{
	  printf ("Line %i to long (more than %i chars).\n", linenr, BUFSIZE);
	  fclose (file);
	  return NULL;
	}
      if (count != 0)
	{
#ifdef DEBUG
	  printf ("Line: '%s'\n", line);
#endif
	}

      /* skip comments and empty lines */
      if ((line[0] == '#') || (count == 0))
	{
	  continue;
	}

      /* XXX TODO could eliminate this check */
      if (count != 0)
	{
	  if (strncmp ((char *)line, "count=", 6) == 0)
	    {
	      /* read out the count to alloc memory */
	      /* 9999 diff. charsets should be enough for everyone... */
	      strncpy ((char *)buffer,(const char *) &line[6], 4);
	      buffer[4] = 0;
	      MAXSETS = atoi ((const char *)buffer);
	      if (MAXSETS == 0)
		{
		  printf ("At least one charset should exist!\n");
		  fclose (file);
		  return NULL;
		}
#ifdef DEBUG
	      printf ("count = %i\n", MAXSETS);
#endif
	    }

	  if ((line[0] == '0') || (line[0] == '1') || (line[0] == '2') ||
	      (line[0] == '3'))
	    {
	      /* start of charset definitions, so make sure we have memory */
	      if ((MAXSETS == 0) && (alloc != 0))
		{
		  printf ("Charset count must come before definitions!\n");
		  fclose (file);
		  return NULL;
		}
	      if ((MAXSETS == 0) && (alloc == 0))
		{
		  printf ("At least one charset should exist!\n");
		  fclose (file);
		  return NULL;
		}
	      if (alloc == 0)
		{
#ifdef DEBUG
		  printf ("malloc for %i charsets\n", MAXSETS);
#endif
		  charset = malloc (sizeof (struct ssCharset) * MAXSETS);
		  if (charset == NULL)
		    {
		      printf ("Cannot alloc memory!\n");
		      fclose (file);
		      return NULL;
		    }
		  for (j = 0; j < MAXSETS; j++)
		    {
		      charset[j].id = 0;	/* not yet read */
		    }
		  alloc = 1;	/* flag that we already allocated */
		}
#ifdef DEBUG
	      printf ("parsing now line %i\n", linenr);
#endif
	      /* now parse the charset definition line */
	      if (line[1] != ':')
		{
		  printf
		    ("Error: Charset type must be followed by ':' in line %i\n",
		     linenr);
		  fclose (file);
		  return NULL;
		}
	      charset[set].type = SET_SIMPLE;
	      if (line[0] == '1')
		{
		  charset[set].type = SET_GROUPED;
		}
	      if (line[0] == '2')
		{
		  charset[set].type = SET_DICTIONARY;
		}
	      if (line[0] == '3')
		{
		  charset[set].type = SET_EXTRACT;
		}
	      pos = 2;		/* skip /^(0|1|2|3): */
	      /* convert first the id of the set */
	      if (line[pos] == ':')
		{
		  printf ("Error: Charset ID must not be empty in line %i\n",
			  linenr);
		  fclose (file);
		  return NULL;
		}
	      i = 0;
	      while (pos < count && line[pos] != ':' && i <= 4)
		{
		  buffer[i++] = line[pos++];
		  buffer[i] = 0;
		}
	      if (i > 4)
		{
		  printf ("Error: Charset ID to big in line %i\n", linenr);
		  fclose (file);
		  return NULL;
		}
	      convert = atoi ((const char *)buffer);
	      if (convert <= 0)
		{
		  printf
		    ("Error: Charset ID must be > 0, but is %i, in line %i\n",
		     convert, linenr);
		  fclose (file);
		  return NULL;
		}
	      /* test that the charset number in convert is one of the existing ones */
	      i = 0;
	      while (i < MAXSETS)
		{
		  /* only for read in ones, since unread ones have still id==0
		     convert is >0 here, this is tested above */
		  if (charset[i].id == (unsigned int)convert)
		    {
		      printf
			("Error: Charset ID %i defined twice in line %i\n",
			 convert, linenr);
		      fclose (file);
		      return NULL;
		    }
		  i++;
		}
	      charset[set].id = convert;

#ifdef DEBUG
	      printf ("Reading charset %i ... ", charset[set].id);
#endif
	      /* now read either the chars or the pos=set fields */
	      charset[set].length = 0;	/* start anew (simple) */
	      charset[set].def = NULL;	/* clear def (grouped) */
	      charset[set].paircnt = 0;	/* also current pair (grouped) */
	      while (pos < count)
		{
		  pos++;	/* skip the ':' separator */
		  if (pos == count)
		    {
		      /* pos == count => trailing ':' (or garbage, but we don't look) */
		      break;
		    }
		  /* convert the characters */
		  i = 0;
		  while (pos < count && line[pos] != ':' && i <= 8)
		    {
		      buffer[i++] = line[pos++];
		    }
		  buffer[i] = 0;
		  if (i > MAXFIELDSIZE)
		    {
		      printf ("Error: Field to big in line %i\n", linenr);
		      fclose (file);
		      return NULL;
		    }
		  if (charset[set].type == SET_SIMPLE)
		    {
		    /* convert hex to dec */
		    idx = i - 1;
		    convert = 0;
		    mul = 1;
		    while (idx >= 0)
		      {
		      /* assumes some ASCIIism */
		      if (buffer[idx] >= '0' && buffer[idx] <= '9')
			{
			convert += mul * (buffer[idx] - '0');
			}
		      else
			{
			if (buffer[idx] > 'f')
			  {
			  buffer[idx] -= 'A';
			  }	/* to lower */
			if (buffer[idx] < 'a' || buffer[idx] > 'f')
			  {
			  printf ("Error: Illegal hex character %c in line %i\n",
			     buffer[idx], linenr);
			  fclose (file);
			  return NULL;
			  }
			convert += mul * (buffer[idx] - 'a' + 10);
			}
			mul *= 16;
			idx--;
		      }	/* for each nibble */
		      /* got one, so put it in and add one more to len */
		      charset[set].chars[charset[set].length++] = convert;
		    }
		  if (charset[set].type == SET_GROUPED)
		    {
		      /* split pos=id into pos and id at = */
		      j = 0;
		      convert = 0;
		      mul = 1;
		      while (j < i)
			{
			  if (buffer[j] == '=')
			    {
			      break;
			    }
			  if ((buffer[j] != '-')
			      && (buffer[j] < '0' || buffer[j] > '9'))
			    {
			      printf
				("Error: Illegal character '%c' in line %i\n",
				 buffer[j], linenr);
			      fclose (file);
			      return NULL;
			    }
			  temp[j] = buffer[j];
			  j++;
			}
		      temp[j] = 0;
		      if (buffer[j] != '=')
			{
			  printf
			    ("Error: Illegal character '%c' (expect: '=') in line %i\n",
			     buffer[j], linenr);
			}
		      convert = atoi ((const char *)temp);	/* position */
#ifdef DEBUG
		      printf ("convert = %i (from %s)\n", convert, temp);
#endif
		      if (convert != 0)	/* not default position? */
			{
			  charset[set].pos[charset[set].paircnt] = convert;
			}
		      j++;
		      k = 0;	/* skip = */
		      while (j < i)
			{
			  if ((buffer[j] < '0' || buffer[j] > '9'))
			    {
			      printf
				("Error: Illegal character %c (expect: 0..9) in line %i\n",
				 buffer[j], linenr);
			      fclose (file);
			      return NULL;
			    }
			  temp[k++] = buffer[j++];
			}
		      temp[k] = 0;
		      j = atoi ((const char *)temp);
#ifdef DEBUG
		      printf ("j = %i (from %s)\n", j, temp);
#endif
		      /* check if simple set for this pos is valid */
		      charset[set].set[charset[set].paircnt] = NULL;
		      i = 0;
		      while (i < MAXSETS)
			{
			  /* only for read in ones, since unread ones have still id==0 */
			  if (charset[i].id == j)
			    {
			      if (charset[i].type != SET_SIMPLE)
				{
				  printf
				    ("Error: Set %i referenced in set %i in line %i must be a simple set, not another grouped or dictionary one.\n",
				     i, set, linenr);
				  fclose (file);
				  return NULL;
				}
			      /* store pointer to set in set[] or def, depending on */
			      if (convert != 0)
				{
				  charset[set].set[charset[set].paircnt] =
				    &charset[i];
				}
			      else
				{
				  charset[set].def = &charset[i];
				}
			      break;
			    }
			  i++;
			}
		      if (i >= MAXSETS)
			{
			  printf
			    ("Error: Simple set with id %i not found in line %i\n",
			     j, linenr);
			  fclose (file);
			  return NULL;
			}

		      if (convert != 0)	/* not for default one */
			{
			  charset[set].paircnt++;	/* next pair */
			  if (charset[set].paircnt > MAXPAIRS)	/* oups */
			    {
			      printf
				("Error: Too many pairs in grouped set in line %i. Recompile with bigger table.\n",
				 linenr);
			      fclose (file);
			      return NULL;
			    }
			}
		    }		/* for GROUPED one */
		  if (charset[set].type == SET_DICTIONARY)
		    {
#ifdef DEBUG
		      printf ("DEBUG: dictionary set found (field len %i)\n",
			      i);
#endif
		      /* format is: position,charset_id,startlen,endlen 
		         position is either 0 (append) or 1 (prepend)     */
		      j = 0;
		      s = 0;
		      stage = 0;	/* how I hate text parsing in C... */
		      while (j < i)	/* i = strlen(value) */
			{
			  temp[s++] = buffer[j];
			  if (buffer[j] == ',')
			    {
			      if (stage >= 3)
				{
				  printf
				    ("  Error: '%s' contains more than 4 values.\n",
				     buffer);
				  fclose (file);
				  return NULL;
				}
			      temp[s] = 0;	/* zero terminate */
			      s = 0;	/* restart in temp */
			      /* printf ("Split %s\n",temp); */
			      temp_values[stage++] = atoi ((const char *)temp);
			    }
			  j++;
			}
		      /* convert last value, too */
		      temp_values[stage] = atoi ((const char *)temp);
#ifdef DEBUG
		      printf ("Decoded %i %i %i %i\n",
			      temp_values[0], temp_values[1],
			      temp_values[2], temp_values[3]);
#endif
		      /* check that the given setid is valid */
		      charset[set].set[charset[set].paircnt] = NULL;
		      i = 0;
		      while (i < MAXSETS)
			{
			  /* only for read in ones, since unread ones have still id==0 */
			  if (charset[i].id == temp_values[1])
			    {
			      if ((charset[i].type != SET_SIMPLE) &&
				  (charset[i].type != SET_GROUPED))
				{
				  printf
				    ("Error: Set %i referenced in set %i in line %i must be a simple or grouped set, not another dictionary one.\n",
				     i, set, linenr);
				  fclose (file);
				  return NULL;
				}
			      /* store pointer to set in set[] */
			      charset[set].set[charset[set].paircnt] =
				&charset[i];
			      break;
			    }
			  i++;
			}
		      if (i >= MAXSETS)
			{
			  printf ("Error: Set id %i not found in line %i\n",
				  temp_values[0], linenr);
			  fclose (file);
			  return NULL;
			}
		      /* store read values */
		      s = charset[set].paircnt++;	/* shortcut and inc */
		      charset[set].pos[s] = temp_values[0];
		      charset[set].startlen[s] = temp_values[2];
		      charset[set].endlen[s] = temp_values[3];
		    }		/* for DICTIONARY one */
		}		/* for each field */

#ifdef DEBUG
	      if (charset[set].type == SET_SIMPLE)
		{
		  printf (" done. %x .. %x (%i chars)\n",
			  charset[set].chars[0],
			  charset[set].chars[charset[set].length - 1],
			  charset[set].length);
		}
#endif
	      if (charset[set].type == SET_GROUPED)
		{
		  if (charset[set].def == NULL)
		    {
		      printf
			("Error: Grouped set id %i has no default set in line %i\n",
			 charset[set].id, linenr);
		      fclose (file);
		      return NULL;
		    }
#ifdef DEBUG
		  printf (" done for grouped (pairs: %i)\n",
			  charset[set].paircnt);
#endif
		}
#ifdef DEBUG
	      if (charset[set].type == SET_DICTIONARY)
		{
		  /* SET_DICTIONARY */
		  printf (" done for dictionary (append/prepend sets: %i)\n",
			  charset[set].paircnt);
		}
#endif
	      set++;		/* have one more now */
	    }			/* charset definition line? */
	  /* lines not starting with 'c', '0', '1' or '2' are simple ignored */
	}			/* empty line? */
      else
	{
#ifdef DEBUG
	  printf ("Skipping line %i\n", linenr);
#endif
	}
      if (set > MAXSETS)
	{
	  printf ("Error: More sets defined than declared in line %i\n",
		  linenr);
	  fclose (file);
	  return NULL;
	}
    }				/* while not eof */

  if (set < MAXSETS)
    {
      printf ("Error: Less sets defined than declared at eof.\n");
      fclose (file);
      return NULL;
    }

  printf ("read %i charsets. Closing file.\n", MAXSETS);
  fclose (file);
  return charset;
}

/* ************************************************************************ */
/* INTERNAL: read job/charset description file
   return codes: ptr to charsets - okay
		 NULL 		 - error 	*/

unsigned int
pwdgen_readset (const char *set, struct ssPWD *sPWD)
  {
  unsigned long rc;
  struct ssKey *cfg, *key;
  char valid[256] =
    "password_prefix,prefix,config_encoding,dictionary_file,dictionary_mutations,dictionary_stages,charset_id,input_encoding,output_encoding";

  /* read the config file, if we find it */
  cfg = pwdcfg_read (set);

  if (NULL == cfg)
    {
      return 1;
    }

  /* check that we have only allowed keys, exit() otherwise */
  rc = pwdcfg_valid_keys (cfg, &valid[0], 0);
  if (0 != rc)
    {
      return rc;
    }

  key = pwdcfg_find_key (cfg, "charset_id", CFG_NOFAIL);
  if (NULL != key)
    {
      sPWD->id = atoi (key->value);
    }

  /* set all the others */
  pwdgen_set_cfg_keys (sPWD, cfg);

  /* XXX TODO: free memory from cfg */

  return 0;
  }

/** <b>INTERNAL</b>
    Interpret the following set of keys from a given chunk or job/charset
    description file:\n
    - prefix
    - config_encoding
    - password_prefix (legacy support)
    - dictionary_file
    - dictionary_file_offset
    - dictionary_mutations
    - dictionary_stages
    - input_encoding
    - output_encoding
    */

void
pwdgen_set_cfg_keys (struct ssPWD *sPWD, const struct ssKey *cfg)
  {
  struct ssKey *key;
  struct sPwdString* prefix;

  key = pwdcfg_find_key (cfg, "password_prefix", CFG_OPTIONAL);
  if (NULL != key)
    {
    /* legacy support for prefix in hex */
    h2a(key->value);
    /* prefix is read-only */
    prefix = (struct sPwdString*)sPWD->prefix;
    strncpy (prefix->content, key->value, 250);
    prefix->bytes = strlen(prefix->content);
    prefix->characters = -1;
    prefix->encoding = ISO_8859_1;
    printf ("\n Warning: Found legacy key 'password_prefix', assuming hex and ISO-8859-1.\n");
    fflush(NULL);
    }
  key = pwdcfg_find_key (cfg, "prefix", CFG_OPTIONAL);
  if (NULL != key)
    {
    if (sPWD->prefix->bytes != 0)
      {
      printf (" Warning: Encountered both 'password_prefix' and 'prefix' in config file.\n");
      printf ("          'prefix' will take precedence and overwrite 'password_prefix'!\n");
      }
    /* prefix is read-only */
    prefix = (struct sPwdString*)sPWD->prefix;
    strncpy (prefix->content, key->value, 250);
    prefix->content[250] = 0;			/* zero term. in case */
    prefix->bytes = strlen(prefix->content);
    prefix->characters = -1;
    prefix->encoding = UTF_8;
    key = pwdcfg_find_key (cfg, "config_encoding", CFG_OPTIONAL);
    if (NULL != key)
      {
      prefix->encoding = pwdgen_encoding_from_name(key->value);
      if (prefix->encoding == INVALID_ENCODING)
	{
        printf (" Error: 'config_encoding=%s' is invalid.\n", key->value);
	pwdgen_error(sPWD, -1);
	return;
	}
      }
    else
      {
      printf (" Warning: 'config_encoding' not set, assuming UTF-8.\n");
      }
    }

  key = pwdcfg_find_key (cfg, "dictionary_file", CFG_OPTIONAL);
  if (NULL != key)
    {
    strncpy (sPWD->dictionary, key->value, 250);
    sPWD->dictionary[250] = 0;
    }

  /* allow dictionary as alias for dictionary_file */
  key = pwdcfg_find_key (cfg, "dictionary", CFG_OPTIONAL);
  if (NULL != key)
    {
    strncpy (sPWD->dictionary, key->value, 250);
    sPWD->dictionary[250] = 0;
    }

  /* this one will only be used from a chunk description file, not from
     a job description file, but we handle both branches here */
  key = pwdcfg_find_key (cfg, "dictionary_file_offset", CFG_OPTIONAL);
  if (NULL != key)
    {
    sPWD->dict_file_ofs = atol (key->value);
    }

  /* don't use as_int() because it warns needlessly */
  key = pwdcfg_find_key (cfg, "dictionary_mutations", CFG_OPTIONAL);
  if (NULL != key)
    {
    sPWD->todo_mutations = atoi (key->value);
    }

  key = pwdcfg_find_key (cfg,"dictionary_stages", CFG_OPTIONAL);
  if (NULL != key)
    {
    sPWD->todo_stages = atoi (key->value);
    }

  key = pwdcfg_find_key (cfg,"input_encoding", CFG_OPTIONAL);
  if (NULL != key)
    {
    sPWD->generator_encoding = pwdgen_encoding_from_name(key->value);
    if (sPWD->generator_encoding == INVALID_ENCODING)
      {
      printf (" Error: input_encoding '%s' is not a valid encoding.\n",
        key->value);
      pwdgen_error(sPWD, -1);
      return;
      }
    }

  /* the default is no encodings set, so that pwdgen_init() will set a default */
  sPWD->encodings = 0;

  /* after we have set some default encodings, see if the config file wants
     to override them */
  key = pwdcfg_find_key (cfg,"output_encoding", CFG_OPTIONAL);
  if (NULL != key)
    {
    /* XXX TODO: handle more than one ("UTF-8, ISO-8859-1") and also errors */
    printf (" Config requests to produce: %s\n", key->value);
    pwdgen_use_encoding(sPWD, pwdgen_encoding_from_name(key->value));
    }

  }

